File: max_sub_array.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int max_sub_array(int *arr, int sz) {
  int i;
  int max = INT_MIN;
  int val = 0;
  for (i = 0; i < sz; i++) {
    int tval = val + arr[i];
    if (tval > max) {
      max = tval;
    }
    if (tval > 0) {
      val = val + arr[i];
    } else {
      val = 0;
    }
  }
  return max;
}

int main(int argc, char **argv)
{
  int i;
  int x[argc-1];
  for (i = 1; i < argc; i++) {
    x[i-1] = atoi(argv[i]);
  }

  printf("%d\n", max_sub_array(x, argc-1));
}

File: haskell/maxSubArray.hs
maxSubArray' :: (Num a, Ord a) => [a] -> a -> a -> a
maxSubArray' [] v m = m
maxSubArray' (x:y) v m
  | (x + v) <= 0 = maxSubArray' y 0 m
  | (x + v) > 0 = maxSubArray' y (x + v) (max (x + v) m)


maxSubArray :: (Num a, Ord a) => [a] -> a
maxSubArray (x:y) = maxSubArray' y 0 x
